snuba/utils/streams/kafka.py

```diff
@@ -194,14 +214,52 @@ def poll(self, timeout: Optional[float] = None) -> Optional[KafkaMessage]:
             else:
                 raise ConsumerError(str(error))
 
-        return KafkaMessage(
+        result = KafkaMessage(
             TopicPartition(message.topic(), message.partition()),
             message.offset(),
             message.value(),
         )
 
+        self.__offsets[result.stream] = result.offset + 1
+
+        return result
+
+    def tell(self) -> Mapping[TopicPartition, int]:
+        if self.__state in {KafkaConsumerState.CLOSED, KafkaConsumerState.ERROR}:
+            raise InvalidState(self.__state)
+
+        return self.__offsets
+
+    def __seek(self, offsets: Mapping[TopicPartition, int]) -> None:
+        if self.__state is KafkaConsumerState.ASSIGNING:
+            # Trying to use ``seek`` while in an assignment callback will yield
+            # throw an "Erroneous state" error, so seeking actually happens via
+            # assignment in the callback.
```

Comment:
As long as the system is not inherently multithreading (assignment happening on a separate thread) I don't think we need to synchronize everything. I would mention that in the documentation though "this is not threadsafe" or something like that.
I would have been more concerned if assignment happened in a separate thread because there the callsite would not have been able to add synchronization.